src/minuteman_mesos_poller.erl

```diff
@@ -212,28 +223,45 @@ handle_poll_failure(State = #state{last_poll_time = LastPollTime}) ->
     handle_poll_failure(TimeSinceLastPoll, State).
 
 handle_poll_failure(TimeSinceLastPoll, State) when TimeSinceLastPoll > ?AGENT_TIMEOUT_SECS ->
-    VIPBEs = [],
-    LashupVIPs = lashup_kv:value(?VIPS_KEY),
-    Ops = generate_ops(State#state.agent_ip, VIPBEs, LashupVIPs),
-    maybe_perform_ops(Ops),
-    State;
+    handle_poll_changes([], State);
 handle_poll_failure(_TimeSinceLastPoll, State) ->
     State.
 
 -spec(handle_poll_state(mesos_state_client:mesos_agent_state(), state()) -> state()).
-handle_poll_state(MesosState, State) ->
-    VIPBEs = collect_vips(MesosState, State),
+handle_poll_state(MesosState, State0) ->
+    VIPBEs = collect_vips(MesosState, State0),
+    FlatVIPBEs = flatten_vips(VIPBEs),
+    State1 = handle_poll_changes(FlatVIPBEs, State0),
+    State1#state{last_poll_time = erlang:monotonic_time()}.
+
+-spec(handle_poll_changes([vip_be()], state()) -> state()).
+handle_poll_changes(FlatVIPBEs, State) ->
+    AgentIP = State#state.agent_ip,
+    AgentBEIPs = State#state.backend_ips,
     LashupVIPs = lashup_kv:value(?VIPS_KEY),
-    Ops = generate_ops(State#state.agent_ip, VIPBEs, LashupVIPs),
+    Ops = generate_ops(AgentBEIPs, FlatVIPBEs, LashupVIPs),
     maybe_perform_ops(Ops),
-    State#state{last_poll_time = erlang:monotonic_time()}.
+    NewBEIPs = maybe_update_bes(AgentIP, AgentBEIPs, FlatVIPBEs),
```

Comment:
It would be nicer / more readable if `maybe_update_bes` were split into 2 functions / responsibilities. Currently it is calculating the new "BEs" as well as interacting with dets. A better pattern might be:

```
...
NewBEIPs = beips_from_records(FlatVIPBEs),
case maybe_update_bes(AgentIP, AgentBEIPs, NewBEIPs) of
    ok -> State#state{backend_ips = NewBEIPs};
    _Error -> State %% Maybe also log the error here, and/or fail hard because dets was misbehaving!
end.

beips_from_records(FlatVIPBEs) ->
    Ips = [BE || #vip_be{backend_ip = BE} <- FlatVIPBEs],
    ordsets:from_list(Ips).

maybe_update_bes(_, OldBEs, OldBEs) ->
    ok;
maybe_update_bes(AgentIP, _, NewBEs) ->
    case dets:insert(?AGENT_BE, {AgentIP, NewBEs}) of
        ok -> dets:sync(?AGENT_BE);
        Error -> Error
    end.
```